drivers/input/touchscreen/himax.c

```diff
@@ -0,0 +1,462 @@
+	if (!ts->input_dev) {
+		dev_err(&ts->client->dev,
+			"%s: Failed to allocate input device\n", __func__);
+		return -ENOMEM;
+	}
+
+	ts->input_dev->name = "Himax Touchscreen";
+
+	input_set_capability(ts->input_dev, EV_ABS, ABS_MT_POSITION_X);
+	input_set_capability(ts->input_dev, EV_ABS, ABS_MT_POSITION_Y);
+	input_set_abs_params(ts->input_dev, ABS_MT_WIDTH_MAJOR, 0, 200, 0, 0);
+	input_set_abs_params(ts->input_dev, ABS_MT_TOUCH_MAJOR, 0, 200, 0, 0);
+
+	touchscreen_parse_properties(ts->input_dev, true, &ts->props);
+
+	error = input_mt_init_slots(ts->input_dev, HIMAX_MAX_POINTS,
+				    INPUT_MT_DIRECT | INPUT_MT_DROP_UNUSED);
+	if (error) {
+		dev_err(&ts->client->dev, "Failed to initialize MT slots: %d",
+			error);
+		return error;
+	}
+
+	error = input_register_device(ts->input_dev);
+	if (error) {
+		dev_err(&ts->client->dev, "Failed to register input device: %d",
+			error);
+		return error;
+	}
+
+	return 0;
+}
+
+static u8 himax_event_get_num_points(const struct himax_event *event)
+{
+	if (event->num_points == 0xff)
+		return 0;
+	else
+		return event->num_points & 0x0f;
+}
+
+static bool himax_event_point_is_valid(const struct himax_event_point *point)
+{
+	return point->x != 0xffff && point->y != 0xffff;
+}
+
+static u16 himax_event_point_get_x(const struct himax_event_point *point)
+{
+	return be16_to_cpu(point->x);
+}
+
+static u16 himax_event_point_get_y(const struct himax_event_point *point)
+{
+	return be16_to_cpu(point->y);
+}
+
+static bool himax_process_event_point(struct himax_ts_data *ts,
+				      const struct himax_event *event,
+				      int point_index)
+{
+	const struct himax_event_point *point = &event->points[point_index];
+	u16 x = himax_event_point_get_x(point);
+	u16 y = himax_event_point_get_y(point);
+	u8 w = event->majors[point_index];
+
+	if (himax_event_point_is_valid(point)) {
+		input_mt_slot(ts->input_dev, point_index);
+		input_mt_report_slot_state(ts->input_dev, MT_TOOL_FINGER, true);
+		touchscreen_report_pos(ts->input_dev, &ts->props, x, y, true);
+		input_report_abs(ts->input_dev, ABS_MT_TOUCH_MAJOR, w);
+		input_report_abs(ts->input_dev, ABS_MT_WIDTH_MAJOR, w);
+		return true;
+	} else {
+		return false;
+	}
+}
+
+static void himax_process_event(struct himax_ts_data *ts,
+				const struct himax_event *event)
+{
+	int i;
+	int num_points_left = himax_event_get_num_points(event);
+
+	for (i = 0; i < HIMAX_MAX_POINTS && num_points_left > 0; i++) {
+		if (himax_process_event_point(ts, event, i))
+			num_points_left--;
+	}
+
+	input_mt_sync_frame(ts->input_dev);
+	input_sync(ts->input_dev);
+}
+
+static bool himax_verify_checksum(struct himax_ts_data *ts,
+				  const struct himax_event *event)
+{
+	u8 *data = (u8 *)event;
+	int i;
+	u16 checksum = 0;
+
+	for (i = 0; i < sizeof(*event); i++)
+		checksum += data[i];
+
+	if ((checksum & 0x00ff) != 0) {
+		dev_err(&ts->client->dev, "Wrong event checksum: %04x\n",
+			checksum);
+		return false;
+	} else {
+		return true;
+	}
```

Comment:
I don't think you see many of these trailing `else` blocks upstream. Feel prepared to get some feedback about it on the mailing list ;)